#**BOSTON DATASET DATA ANALYSIS**  

IMPORTING THE LIBRARIES

In [37]:
import numpy as np
import pandas as pd
import sklearn

In [7]:
from sklearn.datasets import load_boston
boston_df=load_boston()

GETTING THE DATA/READING DATA

* **Boston dataset is a dictionary which holds keys and values for each key.**
* Let's take the look at the keys

In [8]:
boston_df.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [10]:
#Looking at the shape of the dataset

print(boston_df.data.shape)

(506, 13)


Hence, there are 506 rows of data and 13 columns.

In [15]:
#Feature names of the columns

print(boston_df.feature_names)

['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


In [13]:
#Dataset info
print(boston_df.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

CONVERSION OF DATASET INTO PANDAS DATAFRAME

In [16]:
boston_pd=pd.DataFrame(boston_df.data,columns=boston_df.feature_names)

#printing first 10 rows
boston_pd.head(10)                    

CRIM    ZN  INDUS  CHAS    NOX  ...  RAD    TAX  PTRATIO       B  LSTAT
0  0.00632  18.0   2.31   0.0  0.538  ...  1.0  296.0     15.3  396.90   4.98
1  0.02731   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  396.90   9.14
2  0.02729   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  392.83   4.03
3  0.03237   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  394.63   2.94
4  0.06905   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  396.90   5.33
5  0.02985   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  394.12   5.21
6  0.08829  12.5   7.87   0.0  0.524  ...  5.0  311.0     15.2  395.60  12.43
7  0.14455  12.5   7.87   0.0  0.524  ...  5.0  311.0     15.2  396.90  19.15
8  0.21124  12.5   7.87   0.0  0.524  ...  5.0  311.0     15.2  386.63  29.93
9  0.17004  12.5   7.87   0.0  0.524  ...  5.0  311.0     15.2  386.71  17.10

[10 rows x 13 columns]

In [17]:
#printing last 10 rows

boston_pd.tail(10)

CRIM   ZN  INDUS  CHAS    NOX  ...  RAD    TAX  PTRATIO       B  LSTAT
496  0.28960  0.0   9.69   0.0  0.585  ...  6.0  391.0     19.2  396.90  21.14
497  0.26838  0.0   9.69   0.0  0.585  ...  6.0  391.0     19.2  396.90  14.10
498  0.23912  0.0   9.69   0.0  0.585  ...  6.0  391.0     19.2  396.90  12.92
499  0.17783  0.0   9.69   0.0  0.585  ...  6.0  391.0     19.2  395.77  15.10
500  0.22438  0.0   9.69   0.0  0.585  ...  6.0  391.0     19.2  396.90  14.33
501  0.06263  0.0  11.93   0.0  0.573  ...  1.0  273.0     21.0  391.99   9.67
502  0.04527  0.0  11.93   0.0  0.573  ...  1.0  273.0     21.0  396.90   9.08
503  0.06076  0.0  11.93   0.0  0.573  ...  1.0  273.0     21.0  396.90   5.64
504  0.10959  0.0  11.93   0.0  0.573  ...  1.0  273.0     21.0  393.45   6.48
505  0.04741  0.0  11.93   0.0  0.573  ...  1.0  273.0     21.0  396.90   7.88

[10 rows x 13 columns]

ADDING A NEW COLUMN
* We are adding a new column of the target values to the dataframe.

In [18]:
boston_pd['Median']=boston_df.target
boston_pd.head()

CRIM    ZN  INDUS  CHAS    NOX  ...    TAX  PTRATIO       B  LSTAT  Median
0  0.00632  18.0   2.31   0.0  0.538  ...  296.0     15.3  396.90   4.98    24.0
1  0.02731   0.0   7.07   0.0  0.469  ...  242.0     17.8  396.90   9.14    21.6
2  0.02729   0.0   7.07   0.0  0.469  ...  242.0     17.8  392.83   4.03    34.7
3  0.03237   0.0   2.18   0.0  0.458  ...  222.0     18.7  394.63   2.94    33.4
4  0.06905   0.0   2.18   0.0  0.458  ...  222.0     18.7  396.90   5.33    36.2

[5 rows x 14 columns]

DEALING WITH NULL VALUES

In [20]:
boston_pd.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
Median     0
dtype: int64

* From the above, we can see that there is no null values in the dataset.
* Let's gets some statistic about the dataset;

In [21]:
boston_pd.describe()

CRIM          ZN       INDUS  ...           B       LSTAT      Median
count  506.000000  506.000000  506.000000  ...  506.000000  506.000000  506.000000
mean     3.613524   11.363636   11.136779  ...  356.674032   12.653063   22.532806
std      8.601545   23.322453    6.860353  ...   91.294864    7.141062    9.197104
min      0.006320    0.000000    0.460000  ...    0.320000    1.730000    5.000000
25%      0.082045    0.000000    5.190000  ...  375.377500    6.950000   17.025000
50%      0.256510    0.000000    9.690000  ...  391.440000   11.360000   21.200000
75%      3.677083   12.500000   18.100000  ...  396.225000   16.955000   25.000000
max     88.976200  100.000000   27.740000  ...  396.900000   37.970000   50.000000

[8 rows x 14 columns]

In [26]:
# Checking new shape of the dataset
print(boston_pd.shape)

(506, 14)


SPLITTING THE DATASET

In [28]:
from sklearn.model_selection import train_test_split

X=boston_pd.drop('Median',axis=1)
Y=boston_pd['Median']

X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.20,random_state=5)

print('X Train shape = ',X_train.shape)
print('X Test shape = ',X_test.shape)
print('Y Train shape = ',Y_train.shape)
print('Y Test shape = ',Y_test.shape)

X Train shape =  (404, 13)
X Test shape =  (102, 13)
Y Train shape =  (404,)
Y Test shape =  (102,)


* Now Importing the Linear Progression Model from sklearn and train it on training set:

In [29]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [33]:
#FITTING MODEL ON THE TRAINING DATASET

linear=LinearRegression()
linear.fit(X_train,Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

PREDICTION AND MODEL EVALUATION FOR THE TRAINING SET

In [35]:
Y_train_predict=linear.predict(X_train)
rsme=(np.sqrt(mean_squared_error(Y_train,Y_train_predict)))

print('The model performance for the training set')
print('Root mean square error is {}'.format(rsme))


The model performance for the training set
Root mean square error is 4.741000992236516


PREDICTION AND MODEL EVALUATION FOR THE TESTING SET

In [36]:
Y_test_predict=linear.predict(X_test)
rsme=(np.sqrt(mean_squared_error(Y_test,Y_test_predict)))

print('The model performance for the test set')
print('Root mean square error is {}'.format(rsme))

The model performance for the test set
Root mean square error is 4.5682920423032


# **CONCLUSION**
* In this notebook,we applied the concepts of Linear Regression on the Boston Housing Dataset.  